service/history/workflow/mutable_state_impl.go

```diff
@@ -6927,13 +6927,20 @@ func (ms *MutableStateImpl) applyUpdatesToStateMachineNodes(
 				// we don't have enough information to recreate all parents
 				return err
 			}
-			newNode, err := parentNode.AddChild(incomingPath[len(incomingPath)-1], nodeMutation.Data)
+
+			key := incomingPath[len(incomingPath)-1]
+			deserilzed, err := parentNode.DeserializeData(key, nodeMutation.Data)
```

Comment:
I would recommend manipulating the parent's internal representation instead of deserializing and doing extra work.

Reply:
updated.

Reply:
updated using parent's internal representation.